File: samples/objc/FirebaseUI-demo-objc/Samples/Auth/FUIAuthViewController.m
//
//  AuthViewController.m
//
//  Copyright (c) 2016 Google Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

@import Firebase;
#import <FirebaseUI/FirebaseUI.h>

#import "FUIAuthViewController.h"
#import "FUIAppDelegate.h"
#import "FUICustomAuthDelegate.h"

#import "FUICustomAuthPickerViewController.h"

NS_ENUM(NSUInteger, UISections) {
  kSectionsSettings = 0,
  kSectionsProviders,
  kSectionsAnonymousSignIn,
  kSectionsName,
  kSectionsEmail,
  kSectionsPhoneNumber,
  kSectionsUID,
  kSectionsAccessToken,
  kSectionsIDToken
};

NS_ENUM(NSUInteger, FIRProviders) {
  kIDPEmail = 0,
  kIDPGoogle,
  kIDPFacebook,
  kIDPTwitter,
  kIDPPhone,
  kIDPAnonymous,
  kIDPMicrosoft,
  kIDPGitHub,
  kIDPYahoo,
  kIDPApple,
};

static NSString *const kFirebaseTermsOfService = @"https://firebase.google.com/terms/";
static NSString *const kFirebasePrivacyPolicy = @"https://firebase.google.com/support/privacy/";

@interface FUIAuthViewController () <FUIAuthDelegate>
@property (weak, nonatomic) IBOutlet UITableViewCell *cellSignIn;
@property (weak, nonatomic) IBOutlet UITableViewCell *cellName;
@property (weak, nonatomic) IBOutlet UITableViewCell *cellEmail;
@property (weak, nonatomic) IBOutlet UISwitch *emailSwitch;
@property (weak, nonatomic) IBOutlet UILabel *emailLabel;
@property (weak, nonatomic) IBOutlet UITableViewCell *cellUID;
@property (weak, nonatomic) IBOutlet UITableViewCell *anonymousSignIn;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *buttonAuthorization;
@property (weak, nonatomic) IBOutlet UITableViewCell *cellAccessToken;
@property (weak, nonatomic) IBOutlet UITableViewCell *cellIdToken;
@property (weak, nonatomic) IBOutlet UITableViewCell *cellPhoneNumber;
@property (weak, nonatomic) IBOutlet UISwitch *customScopeSwitch;

@property (nonatomic) FIRAuth *auth;
@property (nonatomic) FUIAuth *authUI;
// retain customAuthUIDelegate so it can be used when needed
@property (nonatomic) id<FUIAuthDelegate> customAuthUIDelegate;
@property (nonatomic, assign) BOOL isCustomAuthDelegateSelected;

@property (nonatomic) FIRAuthStateDidChangeListenerHandle authStateDidChangeHandle;

@end

@implementation FUIAuthViewController {
  NSInteger _activityCount;
}

#pragma mark - UIViewController methods

- (void)viewDidLoad {
  [super viewDidLoad];

  self.tableView.rowHeight = UITableViewAutomaticDimension;
  self.tableView.estimatedRowHeight = 240;

  self.customAuthUIDelegate = [[FUICustomAuthDelegate alloc] init];

  self.auth = [FIRAuth auth];
  self.authUI = [FUIAuth defaultAuthUI];

  self.authUI.TOSURL = [NSURL URLWithString:kFirebaseTermsOfService];
  self.authUI.privacyPolicyURL = [NSURL URLWithString:kFirebasePrivacyPolicy];

  //set AuthUI Delegate
  [self onAuthUIDelegateChanged:nil];

  //select all Identety providers
  [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:kIDPEmail
                                                          inSection:kSectionsProviders]
                              animated:NO
                        scrollPosition:UITableViewScrollPositionNone];
  [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:kIDPGoogle
                                                          inSection:kSectionsProviders]
                              animated:NO
                        scrollPosition:UITableViewScrollPositionNone];
  [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:kIDPFacebook
                                                          inSection:kSectionsProviders]
                              animated:NO
                        scrollPosition:UITableViewScrollPositionNone];
  [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:kIDPTwitter
                                                          inSection:kSectionsProviders]
                              animated:NO
                        scrollPosition:UITableViewScrollPositionNone];
  [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:kIDPPhone
                                                          inSection:kSectionsProviders]
                              animated:NO
                        scrollPosition:UITableViewScrollPositionNone];
  [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:kIDPAnonymous
                                                          inSection:kSectionsProviders]
                              animated:NO
                        scrollPosition:UITableViewScrollPositionNone];
  [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:kIDPMicrosoft
                                                          inSection:kSectionsProviders]
                              animated:NO
                        scrollPosition:UITableViewScrollPositionNone];
  [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:kIDPGitHub
                                                          inSection:kSectionsProviders]
                              animated:NO
                        scrollPosition:UITableViewScrollPositionNone];
  [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:kIDPYahoo
                                                          inSection:kSectionsProviders]
                              animated:NO
                        scrollPosition:UITableViewScrollPositionNone];
  [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:kIDPApple
                                                          inSection:kSectionsProviders]
                              animated:NO
                        scrollPosition:UITableViewScrollPositionNone];
}

- (void)viewWillAppear:(BOOL)animated {
  [super viewWillAppear:animated];

  __weak FUIAuthViewController *weakSelf = self;
  self.authStateDidChangeHandle =
      [self.auth addAuthStateDidChangeListener:^(FIRAuth *_Nonnull auth, FIRUser *_Nullable user) {
    [weakSelf updateUI:auth withUser:user];
  }];

  self.navigationController.toolbarHidden = NO;
}

-(void)viewWillDisappear:(BOOL)animated {
  [super viewWillDisappear:animated];
  [self.auth removeAuthStateDidChangeListener:self.authStateDidChangeHandle];

  self.navigationController.toolbarHidden = YES;
}

#pragma mark - UITableViewController methods

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
  return UITableViewAutomaticDimension;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  if (indexPath.section == kSectionsAnonymousSignIn && indexPath.row == 0) {
    FIRUser *currentUser = self.authUI.auth.currentUser;
    if (currentUser.isAnonymous) {
      // If the user is anonymous, delete the user to avoid dangling anonymous users.
      if (currentUser.isAnonymous) {
        [currentUser deleteWithCompletion:^(NSError * _Nullable error) {
          if (error) {
            [self showAlertWithTitlte:@"Error" message:error.localizedDescription];
            return;
          }
          [self showAlertWithTitlte:@"" message:@"Anonymous user deleted"];
        }];
      }
      [tableView deselectRowAtIndexPath:indexPath animated:NO];
      return;
    }
    [self signOut];
    [FUIAuth.defaultAuthUI.auth signInAnonymouslyWithCompletion:^(FIRAuthDataResult *_Nullable authResult,
                                                                  NSError *_Nullable error) {
      if (error) {
        NSError *detailedError = error.userInfo[NSUnderlyingErrorKey];
        if (!detailedError) {
          detailedError = error;
        }
        NSLog(@"ERROR: %@", detailedError.localizedDescription);
      }
    }];
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
  }
}

#pragma mark - UI methods

- (void)updateUI:(FIRAuth * _Nonnull) auth withUser:(FIRUser *_Nullable) user {
  if (user) {
    self.cellSignIn.textLabel.text = @"Signed-in";
    self.cellName.textLabel.text = user.displayName;
    self.cellEmail.textLabel.text = user.email;
    self.cellPhoneNumber.textLabel.text = user.phoneNumber;
    self.cellUID.textLabel.text = user.uid;

    // If the user is anonymous, delete the user to avoid dangling anonymous users.
    if (auth.currentUser.isAnonymous) {
      [_anonymousSignIn.textLabel setText:@"Delete Anonymous User"];
    }
    else {
      [_anonymousSignIn.textLabel setText:@"Sign In Anonymously"];
      self.buttonAuthorization.title = @"Sign Out";
    }
  } else {
    [_anonymousSignIn.textLabel setText:@"Sign In Anonymously"];
    self.cellSignIn.textLabel.text = @"Not signed-in";
    self.cellName.textLabel.text = @"";
    self.cellEmail.textLabel.text = @"";
    self.cellPhoneNumber.textLabel.text = @"";
    self.cellUID.textLabel.text = @"";

    self.buttonAuthorization.title = @"Sign In";
  }

  self.cellAccessToken.textLabel.text = [self getAllAccessTokens];
  self.cellIdToken.textLabel.text = [self getAllIdTokens];

  NSArray *selectedRows = [self.tableView indexPathsForSelectedRows];
  [self.tableView reloadData];
  for (NSIndexPath *path in selectedRows) {
    [self.tableView selectRowAtIndexPath:path
                                animated:NO
                          scrollPosition:UITableViewScrollPositionNone];
  }
}
- (IBAction)onAuthUIDelegateChanged:(UISwitch *)sender {
  _isCustomAuthDelegateSelected = sender ? sender.isOn : NO;
  if (_isCustomAuthDelegateSelected) {
    self.authUI.delegate = self.customAuthUIDelegate;
  } else {
    self.authUI.delegate = self;
  }
}

- (IBAction)onAuthorization:(id)sender {
  if (!_auth.currentUser || _auth.currentUser.isAnonymous) {
    FUIAuth.defaultAuthUI.autoUpgradeAnonymousUsers = YES;
    _authUI.providers = [self getListOfIDPs];

    NSString *providerID = self.authUI.providers.firstObject.providerID;
    BOOL isPhoneAuth = [providerID isEqualToString:FIRPhoneAuthProviderID];
    BOOL isEmailAuth = [providerID isEqualToString:FIREmailAuthProviderID];
    BOOL shouldSkipAuthPicker = self.authUI.providers.count == 1 && (isPhoneAuth || isEmailAuth);
    if (shouldSkipAuthPicker) {
      if (isPhoneAuth) {
        FUIPhoneAuth *provider = self.authUI.providers.firstObject;
        [provider signInWithPresentingViewController:self phoneNumber:nil];
      } else if (isEmailAuth) {
        FUIEmailAuth *provider = self.authUI.providers.firstObject;
        [provider signInWithPresentingViewController:self email:nil];
      }
    } else {
      UINavigationController *controller = [self.authUI authViewController];
      if (_isCustomAuthDelegateSelected) {
        controller.navigationBar.hidden = YES;
      }
      [self presentViewController:controller animated:YES completion:nil];
    }
  } else {
    [self signOut];
  }
}

- (IBAction)onEmailSwitchValueChanged:(UISwitch *)sender {
  if (sender.isOn) {
    self.emailLabel.text = @"Password";
  } else {
    self.emailLabel.text = @"Link";
  }
}

#pragma mark - FUIAuthDelegate methods

// this method is called only when FUIAuthViewController is delgate of AuthUI
- (void)authUI:(FUIAuth *)authUI
    didSignInWithAuthDataResult:(nullable FIRAuthDataResult *)authDataResult
                            URL:(nullable NSURL *)url
                          error:(nullable NSError *)error {
  if (error) {
    if (error.code == FUIAuthErrorCodeUserCancelledSignIn) {
      [self showAlertWithTitlte:@"Error" message:error.localizedDescription];
      return;
    }
    if (error.code == FUIAuthErrorCodeMergeConflict) {
      FIRAuthCredential *credential = error.userInfo[FUIAuthCredentialKey];
      [[FUIAuth defaultAuthUI].auth
          signInWithCredential:credential
                    completion:^(FIRAuthDataResult *_Nullable authResult,
                                 NSError *_Nullable error) {
        if (error) {
          [self showAlertWithTitlte:@"Sign-In error" message:error.description];
          NSLog(@"%@",error.description);
          return;
        }
        NSString *anonymousUserID = authUI.auth.currentUser.uid;
        NSString *messsage = [NSString stringWithFormat:@"A merge conflict occurred. The old user"
            " ID was: %@. You are now signed in with the following credential type: %@",
            anonymousUserID, [credential.provider uppercaseString]];
        [self showAlertWithTitlte:@"Merge Conflict" message:messsage];
        NSLog(@"%@", messsage);
      }];
      return;
    }
    NSError *detailedError = error.userInfo[NSUnderlyingErrorKey];
    if (!detailedError) {
      detailedError = error;
    }
    NSLog(@"ERROR: %@", detailedError.localizedDescription);
  }
}

#pragma mark - Helper Methods

- (NSString *)getAllAccessTokens {
  NSMutableString *result = [NSMutableString new];
  for (id<FUIAuthProvider> provider in _authUI.providers) {
    [result appendFormat:@"%@:  %@\n", provider.shortName, provider.accessToken];
  }

  return result;
}

- (NSString *)getAllIdTokens {
  NSMutableString *result = [NSMutableString new];
  for (id<FUIAuthProvider> provider in _authUI.providers) {
    [result appendFormat:@"%@:  %@\n", provider.shortName, provider.idToken];
  }

  return result;
}

- (void)signOut {
  NSError *error;
  [self.authUI signOutWithError:&error];
  if (error) {
    [self showAlertWithTitlte:@"Error" message:error.localizedDescription];
  }
}

- (void)showAlertWithTitlte:(NSString *)title message:(NSString *)message {
  UIAlertController *alert = [UIAlertController alertControllerWithTitle:title
                                                                 message:message
                                                          preferredStyle:UIAlertControllerStyleAlert];
  UIAlertAction* closeButton = [UIAlertAction
                                actionWithTitle:@"Close"
                                style:UIAlertActionStyleDefault
                                handler:nil];
  [alert addAction:closeButton];
  [self presentViewController:alert animated:YES completion:nil];

}

- (NSArray *)getListOfIDPs {
  return [[self class] getListOfIDPs:[self.tableView indexPathsForSelectedRows]
                     useCustomScopes:_customScopeSwitch.isOn
                        useEmailLink:_emailSwitch.isOn];
}

+ (NSArray *)getListOfIDPs:(NSArray<NSIndexPath *> *)selectedRows
           useCustomScopes:(BOOL)useCustomScopes
              useEmailLink:(BOOL)useEmaiLink {
  NSMutableArray *providers = [NSMutableArray new];

  for (NSIndexPath *indexPath in selectedRows) {
    if (indexPath.section == kSectionsProviders) {
      id<FUIAuthProvider> provider;
      switch (indexPath.row) {
        case kIDPEmail:
          if (useEmaiLink) {
            // ActionCodeSettings for email link sign-in.
            FIRActionCodeSettings *actionCodeSettings = [[FIRActionCodeSettings alloc] init];
            actionCodeSettings.URL = [NSURL URLWithString:@"https://fb-sa-1211.appspot.com"];
            actionCodeSettings.handleCodeInApp = YES;
            [actionCodeSettings setAndroidPackageName:@"com.firebase.uidemo"
                                installIfNotAvailable:NO
                                       minimumVersion:@"12"];

            provider = [[FUIEmailAuth alloc] initAuthAuthUI:[FUIAuth defaultAuthUI]
                                               signInMethod:FIREmailLinkAuthSignInMethod
                                            forceSameDevice:NO
                                      allowNewEmailAccounts:YES
                                         requireDisplayName:YES
                                          actionCodeSetting:actionCodeSettings];
          } else {
            provider = [[FUIEmailAuth alloc] initAuthAuthUI:[FUIAuth defaultAuthUI]
                                               signInMethod:FIREmailPasswordAuthSignInMethod
                                            forceSameDevice:NO
                                      allowNewEmailAccounts:YES
                                         requireDisplayName:NO
                                          actionCodeSetting:[[FIRActionCodeSettings alloc] init]];
          }
          break;
        case kIDPGoogle:
          provider = useCustomScopes ? [[FUIGoogleAuth alloc] initWithScopes:@[kGoogleUserInfoEmailScope,
                                                                               kGoogleUserInfoProfileScope,
                                                                               kGoogleGamesScope,
                                                                               kGooglePlusMeScope]]
                                     : [[FUIGoogleAuth alloc] init];
          break;
        case kIDPFacebook:
          provider = useCustomScopes ? [[FUIFacebookAuth alloc] initWithPermissions:@[@"email",
                                                                                      @"user_friends",
                                                                                      @"ads_read"]]
                                     :[[FUIFacebookAuth alloc] init];
          break;
        case kIDPTwitter:
          provider = [FUIOAuth twitterAuthProvider];
          break;
        case kIDPPhone:
          provider = [[FUIPhoneAuth alloc] initWithAuthUI:[FUIAuth defaultAuthUI]];
          break;
        case kIDPAnonymous:
          provider = [[FUIAnonymousAuth alloc] initWithAuthUI:[FUIAuth defaultAuthUI]];
          break;
        case kIDPMicrosoft:
          provider = [FUIOAuth microsoftAuthProvider];
          break;
        case kIDPGitHub:
          provider = [FUIOAuth githubAuthProvider];
          break;
        case kIDPYahoo:
          provider = [FUIOAuth yahooAuthProvider];
          break;
        case kIDPApple:
          provider = [FUIOAuth appleAuthProvider];
          break;
        default:
          break;
      }
      if (provider) {
        [providers addObject:provider];
      }

    }
  }

  return providers;
}

@end

File: samples/swift/FirebaseUI-demo-swift/Samples/Auth/FUIAuthViewController.swift
//
//  Copyright (c) 2016 Google Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import UIKit
import Firebase
import FirebaseUI

let kFirebaseTermsOfService = URL(string: "https://firebase.google.com/terms/")!
let kFirebasePrivacyPolicy = URL(string: "https://firebase.google.com/support/privacy/")!

enum UISections: Int, RawRepresentable {
  case Settings = 0
  case Providers
  case AnonymousSignIn
  case Name
  case Email
  case UID
  case Phone
  case AccessToken
  case IDToken
}

enum Providers: Int, RawRepresentable {
  case Email = 0
  case Google
  case Facebook
  case Twitter
  case Apple
  case Phone
}



/// A view controller displaying a basic sign-in flow using FUIAuth.
class FUIAuthViewController: UITableViewController {
  // Before running this sample, make sure you've correctly configured
  // the appropriate authentication methods in Firebase console. For more
  // info, see the Auth README at ../../FirebaseAuthUI/README.md
  // and https://firebase.google.com/docs/auth/

  fileprivate var authStateDidChangeHandle: AuthStateDidChangeListenerHandle?

  fileprivate(set) var auth: Auth? = Auth.auth()
  fileprivate(set) var authUI: FUIAuth? = FUIAuth.defaultAuthUI()
  fileprivate(set) var customAuthUIDelegate: FUIAuthDelegate = FUICustomAuthDelegate()

  @IBOutlet weak var cellSignedIn: UITableViewCell!
  @IBOutlet weak var cellName: UITableViewCell!
  @IBOutlet weak var cellEmail: UITableViewCell!
  @IBOutlet weak var cellUid: UITableViewCell!
  @IBOutlet weak var cellPhone: UITableViewCell!
  @IBOutlet weak var cellAccessToken: UITableViewCell!
  @IBOutlet weak var cellIdToken: UITableViewCell!
  @IBOutlet weak var cellAnonymousSignIn: UITableViewCell!

  @IBOutlet weak var authorizationButton: UIBarButtonItem!
  @IBOutlet weak var customAuthorizationSwitch: UISwitch!
  @IBOutlet weak var customScopesSwitch: UISwitch!


  override func viewDidLoad() {
    super.viewDidLoad()

    self.authUI?.tosurl = kFirebaseTermsOfService
    self.authUI?.privacyPolicyURL = kFirebasePrivacyPolicy

    self.tableView.selectRow(at: IndexPath(row: Providers.Email.rawValue, section: UISections.Providers.rawValue),
                             animated: false,
                             scrollPosition: .none)
    self.tableView.selectRow(at: IndexPath(row: Providers.Google.rawValue, section: UISections.Providers.rawValue),
                             animated: false,
                             scrollPosition: .none)
    self.tableView.selectRow(at: IndexPath(row: Providers.Facebook.rawValue, section: UISections.Providers.rawValue),
                             animated: false,
                             scrollPosition: .none)
    self.tableView.selectRow(at: IndexPath(row: Providers.Twitter.rawValue, section: UISections.Providers.rawValue),
                             animated: false,
                             scrollPosition: .none)
    self.tableView.selectRow(at: IndexPath(row: Providers.Apple.rawValue, section: UISections.Providers.rawValue),
                             animated: false,
                             scrollPosition: .none)
    self.tableView.selectRow(at: IndexPath(row: Providers.Phone.rawValue, section: UISections.Providers.rawValue),
                             animated: false,
                             scrollPosition: .none)

  }

  override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)

    self.tableView.rowHeight = UITableView.automaticDimension;
    self.tableView.estimatedRowHeight = 240;

    self.authStateDidChangeHandle =
      self.auth?.addStateDidChangeListener(self.updateUI(auth:user:))

    self.navigationController?.isToolbarHidden = false;
  }

  override func viewWillDisappear(_ animated: Bool) {
    super.viewWillDisappear(animated)
    if let handle = self.authStateDidChangeHandle {
      self.auth?.removeStateDidChangeListener(handle)
    }

    self.navigationController?.isToolbarHidden = true;
  }

  override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
    return UITableView.automaticDimension
  }

  override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    if (indexPath.section == UISections.AnonymousSignIn.rawValue && indexPath.row == 0) {
      if (auth?.currentUser?.isAnonymous ?? false) {
        tableView.deselectRow(at: indexPath, animated: false)
        return;
      }
      do {
        try self.authUI?.signOut()
      } catch let error {
        self.ifNoError(error) {}
      }

      auth?.signInAnonymously() { authReuslt, error in
        self.ifNoError(error) {
          self.showAlert(title: "Signed In Anonymously")
        }
      }
      tableView.deselectRow(at: indexPath, animated: false)
    }
  }

 fileprivate func showAlert(title: String, message: String? = "") {
    if #available(iOS 8.0, *) {
      let alertController =
          UIAlertController(title: title, message: message, preferredStyle: .alert)
      alertController.addAction(UIAlertAction(title: "OK",
                                              style: .default,
                                            handler: { (UIAlertAction) in
        alertController.dismiss(animated: true, completion: nil)
      }))
      self.present(alertController, animated: true, completion: nil)
    } else {
      UIAlertView(title: title,
                  message: message ?? "",
                  delegate: nil,
                  cancelButtonTitle: nil,
                  otherButtonTitles: "OK").show()
    }
  }

  private func ifNoError(_ error: Error?, execute: () -> Void) {
    guard error == nil else {
      showAlert(title: "Error", message: error!.localizedDescription)
      return
    }
    execute()
  }

  @IBAction func onAuthorize(_ sender: AnyObject) {
    if (self.auth?.currentUser) != nil {
      if (auth?.currentUser?.isAnonymous != false) {
        auth?.currentUser?.delete() { error in
          self.ifNoError(error) {
            self.showAlert(title: "", message:"The user was properly deleted.")
          }
        }
      } else {
        do {
          try self.authUI?.signOut()
        } catch let error {
          self.ifNoError(error) {
            self.showAlert(title: "Error", message:"The user was properly signed out.")
          }
        }
      }
    } else {
      self.authUI?.delegate = self.customAuthorizationSwitch.isOn ? self.customAuthUIDelegate : nil;

      // If you haven't set up your authentications correctly these buttons
      // will still appear in the UI, but they'll crash the app when tapped.
      self.authUI?.providers = self.getListOfIDPs()

      let providerID = self.authUI?.providers.first?.providerID;
      let isPhoneAuth = providerID == PhoneAuthProviderID;
      let isEmailAuth = providerID == EmailAuthProviderID;
      let shouldSkipAuthPicker = self.authUI?.providers.count == 1 && (isPhoneAuth || isEmailAuth);
      if (shouldSkipAuthPicker) {
        if (isPhoneAuth) {
          let provider = self.authUI?.providers.first as! FUIPhoneAuth;
          provider.signIn(withPresenting: self, phoneNumber: nil);
        } else if (isEmailAuth) {
          let provider = self.authUI?.providers.first as! FUIEmailAuth;
          provider.signIn(withPresenting: self, email: nil);
        }
      } else {
        let controller = self.authUI!.authViewController()
        controller.navigationBar.isHidden = self.customAuthorizationSwitch.isOn
        self.present(controller, animated: true, completion: nil)
      }
    }
  }

  // Boilerplate
  func updateUI(auth: Auth, user: User?) {
    if let user = self.auth?.currentUser {
      self.cellSignedIn.textLabel?.text = "Signed in"
      self.cellName.textLabel?.text = user.displayName ?? "(null)"
      self.cellEmail.textLabel?.text = user.email ?? "(null)"
      self.cellUid.textLabel?.text = user.uid
      self.cellPhone.textLabel?.text = user.phoneNumber

      if (auth.currentUser?.isAnonymous != false) {
        self.authorizationButton.title = "Delete Anonymous User";
      } else {
        self.authorizationButton.title = "Sign Out";
      }
    } else {
      self.cellSignedIn.textLabel?.text = "Not signed in"
      self.cellName.textLabel?.text = "null"
      self.cellEmail.textLabel?.text = "null"
      self.cellUid.textLabel?.text = "null"
      self.cellPhone.textLabel?.text = "null"

      self.authorizationButton.title = "Sign In";
    }

    self.cellAccessToken.textLabel?.text = getAllAccessTokens()
    self.cellIdToken.textLabel?.text = getAllIdTokens()

    let selectedRows = self.tableView.indexPathsForSelectedRows
    self.tableView.reloadData()
    if let rows = selectedRows {
      for path in rows {
        self.tableView.selectRow(at: path, animated: false, scrollPosition: .none)
      }
    }
  }

  func getAllAccessTokens() -> String {
    var result = ""
    for provider in self.authUI!.providers {
      result += (provider.shortName + ": " + (provider.accessToken ?? "null") + "\n")
    }

    return result
  }

  func getAllIdTokens() -> String {
    var result = ""
    for provider in self.authUI!.providers {
      result += (provider.shortName + ": " + (provider.idToken! ?? "null")  + "\n")
    }

    return result
  }

  func getListOfIDPs() -> [FUIAuthProvider] {
    var providers = [FUIAuthProvider]()
    if let selectedRows = self.tableView.indexPathsForSelectedRows {
      for indexPath in selectedRows {
        if indexPath.section == UISections.Providers.rawValue {
          let provider:FUIAuthProvider?

          switch indexPath.row {
          case Providers.Email.rawValue:
            provider = FUIEmailAuth()
          case Providers.Google.rawValue:
            provider = self.customScopesSwitch.isOn ? FUIGoogleAuth(authUI: self.authUI!,
                                                                    scopes: [kGoogleGamesScope,
                                                                             kGooglePlusMeScope,
                                                                             kGoogleUserInfoEmailScope,
                                                                             kGoogleUserInfoProfileScope])
              : FUIGoogleAuth(authUI: self.authUI!)
          case Providers.Twitter.rawValue:
            let buttonColor =
                UIColor(red: 71.0/255.0, green: 154.0/255.0, blue: 234.0/255.0, alpha: 1.0)
            guard let iconPath = Bundle.main.path(forResource: "twtrsymbol", ofType: "png") else {
              NSLog("Warning: Unable to find Twitter icon")
              continue
            }
            provider = FUIOAuth(authUI: self.authUI!,
                                  providerID: "twitter.com",
                                  buttonLabelText: "Sign in with Twitter",
                                  shortName: "Twitter",
                                  buttonColor: buttonColor,
                                  iconImage: UIImage(contentsOfFile: iconPath)!,
                                  scopes: ["user.readwrite"],
                                  customParameters: ["prompt" : "consent"],
                                  loginHintKey: nil)
          case Providers.Facebook.rawValue:
            provider = self.customScopesSwitch.isOn ? FUIFacebookAuth(authUI: self.authUI!,
                                                                      permissions: ["email",
                                                                                    "user_friends",
                                                                                    "ads_read"])
              : FUIFacebookAuth(authUI: self.authUI!)
          case Providers.Apple.rawValue:
            if #available(iOS 13.0, *) {
              provider = FUIOAuth.appleAuthProvider()
            } else {
              provider = nil
            }
          case Providers.Phone.rawValue:
            provider = FUIPhoneAuth(authUI: self.authUI!)
          default: provider = nil
          }

          guard provider != nil else {
            continue
          }

          providers.append(provider!)
        }
      }
    }
    
    return providers
  }

}
